File: demo-app-objc/CloudXObjCRemotePods/Views/Ad/MRECViewController.m
#import "MRECViewController.h"
#import <CloudXCore/CloudXCore.h>
#import "DemoAppLogger.h"
#import "CLXDemoConfigManager.h"
#import "UserDefaultsSettings.h"
#import "NSError+DemoDescription.h"

@interface MRECViewController ()
@property (nonatomic, strong) CLXBannerAdView *mrecAd;
@property (nonatomic, strong) UIButton *autoRefreshButton;
@property (nonatomic, assign) BOOL autoRefreshEnabled;
@property (nonatomic, strong) UserDefaultsSettings *settings;
@end

@implementation MRECViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.autoRefreshEnabled = YES; // Default to enabled
    self.settings = [UserDefaultsSettings sharedSettings];
    
    // Create a vertical stack for buttons
    UIStackView *buttonStack = [[UIStackView alloc] init];
    buttonStack.axis = UILayoutConstraintAxisVertical;
    buttonStack.spacing = 16;
    buttonStack.alignment = UIStackViewAlignmentCenter;
    buttonStack.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:buttonStack];
    
    // Load MREC button (new API — no viewController)
    UIButton *loadButton = [UIButton buttonWithType:UIButtonTypeSystem];
    [loadButton setTitle:@"Load MREC (New API)" forState:UIControlStateNormal];
    [loadButton addTarget:self action:@selector(loadMRECAd) forControlEvents:UIControlEventTouchUpInside];
    loadButton.backgroundColor = [UIColor systemGreenColor];
    [loadButton setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    loadButton.titleLabel.font = [UIFont boldSystemFontOfSize:16];
    loadButton.layer.cornerRadius = 8;
    loadButton.translatesAutoresizingMaskIntoConstraints = NO;
    [buttonStack addArrangedSubview:loadButton];
    
    // Load MREC button (deprecated API — passes viewController)
    UIButton *loadDeprecatedButton = [UIButton buttonWithType:UIButtonTypeSystem];
    [loadDeprecatedButton setTitle:@"Load MREC (Deprecated)" forState:UIControlStateNormal];
    [loadDeprecatedButton addTarget:self action:@selector(loadMRECAdDeprecated) forControlEvents:UIControlEventTouchUpInside];
    loadDeprecatedButton.backgroundColor = [UIColor systemOrangeColor];
    [loadDeprecatedButton setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    loadDeprecatedButton.titleLabel.font = [UIFont boldSystemFontOfSize:16];
    loadDeprecatedButton.layer.cornerRadius = 8;
    loadDeprecatedButton.translatesAutoresizingMaskIntoConstraints = NO;
    [buttonStack addArrangedSubview:loadDeprecatedButton];
    
    // Auto-refresh toggle button (positioned separately above status label)
    self.autoRefreshButton = [UIButton buttonWithType:UIButtonTypeSystem];
    [self.autoRefreshButton setTitle:@"Stop Auto-Refresh" forState:UIControlStateNormal];
    [self.autoRefreshButton addTarget:self action:@selector(toggleAutoRefresh) forControlEvents:UIControlEventTouchUpInside];
    self.autoRefreshButton.backgroundColor = [UIColor systemPurpleColor];
    [self.autoRefreshButton setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    self.autoRefreshButton.titleLabel.font = [UIFont boldSystemFontOfSize:16];
    self.autoRefreshButton.layer.cornerRadius = 8;
    self.autoRefreshButton.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:self.autoRefreshButton];
    
    // Button constraints
    [NSLayoutConstraint activateConstraints:@[
        [buttonStack.centerXAnchor constraintEqualToAnchor:self.view.centerXAnchor],
        [buttonStack.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor constant:100],
        [loadButton.widthAnchor constraintEqualToConstant:200],
        [loadButton.heightAnchor constraintEqualToConstant:44],
        [loadDeprecatedButton.widthAnchor constraintEqualToConstant:200],
        [loadDeprecatedButton.heightAnchor constraintEqualToConstant:44],
        
        // Auto-refresh button positioned above status label
        [self.autoRefreshButton.centerXAnchor constraintEqualToAnchor:self.view.centerXAnchor],
        [self.autoRefreshButton.bottomAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.bottomAnchor constant:-80],
        [self.autoRefreshButton.widthAnchor constraintEqualToConstant:200],
        [self.autoRefreshButton.heightAnchor constraintEqualToConstant:44]
    ]];
    
    // Auto-create and add MREC to view hierarchy immediately
    [self createAndAddMRECToView];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    // Update status based on current ad state
    if (self.mrecAd && !self.isLoading) {
        [self updateStatusUIWithState:AdStateReady];
    } else if (self.isLoading) {
        [self updateStatusUIWithState:AdStateLoading];
    } else {
        [self updateStatusUIWithState:AdStateNoAd];
    }
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    [self resetAdState];
}

- (void)dealloc {
    // Ensure cleanup even if viewWillDisappear wasn't called
    [self resetAdState];
}

- (void)loadMRECAd {
    if (self.isLoading) {
        [self showAlertWithTitle:@"Info" message:@"MREC is already loading."];
        return;
    }
    
    [self resetAdState];
    [self createAndAddMRECToView];
    
    if (!self.mrecAd) {
        return;
    }
    
    self.isLoading = YES;
    [self updateStatusUIWithState:AdStateLoading];
    [self.mrecAd load];
}

- (void)createAndAddMRECToView {
    if (self.mrecAd) return;
    
    NSString *adUnitId = [self adUnitId];
    if (_settings.mrecAdUnitId.length > 0) {
        adUnitId = _settings.mrecAdUnitId;
    }
    self.mrecAd = [[CloudXCore shared] createMRECWithAdUnitId:adUnitId];
    self.mrecAd.delegate = self;
    self.mrecAd.revenueDelegate = self;
    self.mrecAd.placement = @"demo_mrec";
    self.mrecAd.customData = @"screen:detail,section:sidebar";

    if (!self.mrecAd) {
        [self showAlertWithTitle:@"Error" message:@"Failed to create MREC."];
        return;
    }
    
    // Add MREC to view hierarchy immediately
    self.mrecAd.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:self.mrecAd];
    
    [NSLayoutConstraint activateConstraints:@[
        [self.mrecAd.centerXAnchor constraintEqualToAnchor:self.view.centerXAnchor],
        [self.mrecAd.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor constant:210],
        [self.mrecAd.widthAnchor constraintEqualToConstant:300],
        [self.mrecAd.heightAnchor constraintEqualToConstant:250]
    ]];
}

- (void)loadMRECAdDeprecated {
    if (self.isLoading) {
        [self showAlertWithTitle:@"Info" message:@"MREC is already loading."];
        return;
    }
    
    [self resetAdState];
    [self createAndAddMRECToViewDeprecated];
    
    if (!self.mrecAd) {
        return;
    }
    
    self.isLoading = YES;
    [self updateStatusUIWithState:AdStateLoading];
    [self.mrecAd load];
}

#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-declarations"
- (void)createAndAddMRECToViewDeprecated {
    NSString *adUnitId = [self adUnitId];
    if (_settings.mrecAdUnitId.length > 0) {
        adUnitId = _settings.mrecAdUnitId;
    }
    
    self.mrecAd = [[CloudXCore shared] createMRECWithAdUnitId:adUnitId viewController:self];
    self.mrecAd.delegate = self;
    self.mrecAd.revenueDelegate = self;
    self.mrecAd.placement = @"demo_mrec";
    self.mrecAd.customData = @"screen:detail,section:sidebar";

    if (!self.mrecAd) {
        [self showAlertWithTitle:@"Error" message:@"Failed to create MREC (deprecated API)."];
        return;
    }
    
    self.mrecAd.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:self.mrecAd];
    
    [NSLayoutConstraint activateConstraints:@[
        [self.mrecAd.centerXAnchor constraintEqualToAnchor:self.view.centerXAnchor],
        [self.mrecAd.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor constant:210],
        [self.mrecAd.widthAnchor constraintEqualToConstant:300],
        [self.mrecAd.heightAnchor constraintEqualToConstant:250]
    ]];
}
#pragma clang diagnostic pop

// showMRECAd method removed - MREC is auto-added to view on push

- (void)resetAdState {
    if (self.mrecAd) {
        // CRITICAL: Properly destroy the MREC to stop auto-refresh timers and background processing
        [self.mrecAd destroy];
        [self.mrecAd removeFromSuperview];
        self.mrecAd = nil;
    }
    self.isLoading = NO;
}

- (void)toggleAutoRefresh {
    if (!self.mrecAd) {
        return;
    }
    
    self.autoRefreshEnabled = !self.autoRefreshEnabled;
    
    if (self.autoRefreshEnabled) {
        [self.mrecAd startAutoRefresh];
        [self.autoRefreshButton setTitle:@"Stop Auto-Refresh" forState:UIControlStateNormal];
        self.autoRefreshButton.backgroundColor = [UIColor systemRedColor];
    } else {
        [self.mrecAd stopAutoRefresh];
        [self.autoRefreshButton setTitle:@"Start Auto-Refresh" forState:UIControlStateNormal];
        self.autoRefreshButton.backgroundColor = [UIColor systemGreenColor];
    }
}

#pragma mark - CLXBannerDelegate

- (void)didLoadAd:(CLXAd *)ad {
    [[DemoAppLogger sharedInstance] logAdEvent:@"✅ MREC didLoadAd" ad:ad];
    self.isLoading = NO;
    [self updateStatusUIWithState:AdStateReady];
    
    // Don't auto-show - user must press Show MREC button
}

- (void)didFailToLoadAd:(NSString *)adUnitId error:(CLXError *)error {
    // No ad object exists on failure, so use logMessage instead of logAdEvent
    [[DemoAppLogger sharedInstance] logMessage:[NSString stringWithFormat:@"❌ MREC failed to load (%@) - Error: %@", adUnitId, error ? error.localizedDescription : @"Unknown error"]];
    self.isLoading = NO;
    [self updateStatusUIWithState:AdStateNoAd];
    
    dispatch_async(dispatch_get_main_queue(), ^{
        NSString *errorMessage = error ? [error detailedDemoDescription] : @"Unknown error occurred";
        [self showAlertWithTitle:@"MREC Ad Load Failed" message:errorMessage];
    });
}

- (void)didClickAd:(CLXAd *)ad {
    [[DemoAppLogger sharedInstance] logAdEvent:@"👆 MREC didClickAd" ad:ad];
}

- (void)didPayRevenueForAd:(CLXAd *)ad {
    [[DemoAppLogger sharedInstance] logAdEvent:@"💰 MREC didPayRevenueForAd" ad:ad];
}

// Banner-specific delegate methods (MREC is a banner type)
- (void)didExpandAd:(CLXAd *)ad {
    [[DemoAppLogger sharedInstance] logAdEvent:@"🔍 MREC didExpandAd" ad:ad];
    
    dispatch_async(dispatch_get_main_queue(), ^{
        [self showAlertWithTitle:@"MREC Expanded!" 
                         message:@"MREC ad expanded to full screen."];
    });
}

- (void)didCollapseAd:(CLXAd *)ad {
    [[DemoAppLogger sharedInstance] logAdEvent:@"🔽 MREC didCollapseAd" ad:ad];
    
    dispatch_async(dispatch_get_main_queue(), ^{
        [self showAlertWithTitle:@"MREC Collapsed!" 
                         message:@"MREC ad collapsed from full screen."];
    });
}

- (NSString *)adUnitId {
    return [[CLXDemoConfigManager sharedManager] currentConfig].mrecAdUnitId;
}

- (void)loadMREC {
    
    if (self.isLoading || self.mrecAd) {
        return;
    }

    self.isLoading = YES;
    [self updateStatusUIWithState:AdStateLoading];

    NSString *adUnitId = [self adUnitId];
    self.mrecAd = [[CloudXCore shared] createMRECWithAdUnitId:adUnitId];
    self.mrecAd.delegate = self;
    self.mrecAd.revenueDelegate = self;
    self.mrecAd.placement = @"demo_mrec";
    self.mrecAd.customData = @"screen:detail,section:sidebar";

    if (self.mrecAd) {
        [self.mrecAd load];
    } else {
        self.isLoading = NO;
        [self updateStatusUIWithState:AdStateNoAd];
        [self showAlertWithTitle:@"Error" message:@"Failed to create MREC."];
    }
}

@end 

File: demo-app-swift/CloudXSwiftRemotePods/Views/Ad/MRECViewController.swift
import UIKit
import CloudXCore

class MRECViewController: BaseAdViewController, CLXBannerDelegate, CLXAdRevenueDelegate {
    
    private var mrecAd: CLXBannerAdView?
    private var autoRefreshButton: UIButton!
    private var autoRefreshEnabled = true // Default to enabled
    private let settings = UserDefaultsSettings.shared
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.title = "MREC"
        
        // Create a vertical stack for buttons
        let buttonStack = UIStackView()
        buttonStack.axis = .vertical
        buttonStack.spacing = 16
        buttonStack.alignment = .center
        buttonStack.translatesAutoresizingMaskIntoConstraints = false
        view.addSubview(buttonStack)
        
        // Load MREC button (new API — no viewController)
        let loadButton = UIButton(type: .system)
        loadButton.setTitle("Load MREC (New API)", for: .normal)
        loadButton.addTarget(self, action: #selector(loadMRECAd), for: .touchUpInside)
        loadButton.backgroundColor = .systemGreen
        loadButton.setTitleColor(.white, for: .normal)
        loadButton.titleLabel?.font = .boldSystemFont(ofSize: 16)
        loadButton.layer.cornerRadius = 8
        loadButton.translatesAutoresizingMaskIntoConstraints = false
        buttonStack.addArrangedSubview(loadButton)
        
        // Load MREC button (deprecated API — passes viewController)
        let loadDeprecatedButton = UIButton(type: .system)
        loadDeprecatedButton.setTitle("Load MREC (Deprecated)", for: .normal)
        loadDeprecatedButton.addTarget(self, action: #selector(loadMRECAdDeprecated), for: .touchUpInside)
        loadDeprecatedButton.backgroundColor = .systemOrange
        loadDeprecatedButton.setTitleColor(.white, for: .normal)
        loadDeprecatedButton.titleLabel?.font = .boldSystemFont(ofSize: 16)
        loadDeprecatedButton.layer.cornerRadius = 8
        loadDeprecatedButton.translatesAutoresizingMaskIntoConstraints = false
        buttonStack.addArrangedSubview(loadDeprecatedButton)
        
        // Auto-refresh toggle button (positioned separately above status label)
        autoRefreshButton = UIButton(type: .system)
        autoRefreshButton.setTitle("Stop Auto-Refresh", for: .normal)
        autoRefreshButton.addTarget(self, action: #selector(toggleAutoRefresh), for: .touchUpInside)
        autoRefreshButton.backgroundColor = .systemPurple
        autoRefreshButton.setTitleColor(.white, for: .normal)
        autoRefreshButton.titleLabel?.font = .boldSystemFont(ofSize: 16)
        autoRefreshButton.layer.cornerRadius = 8
        autoRefreshButton.translatesAutoresizingMaskIntoConstraints = false
        view.addSubview(autoRefreshButton)
        
        // Button constraints
        NSLayoutConstraint.activate([
            buttonStack.centerXAnchor.constraint(equalTo: view.centerXAnchor),
            buttonStack.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 100),
            loadButton.widthAnchor.constraint(equalToConstant: 200),
            loadButton.heightAnchor.constraint(equalToConstant: 44),
            loadDeprecatedButton.widthAnchor.constraint(equalToConstant: 200),
            loadDeprecatedButton.heightAnchor.constraint(equalToConstant: 44),
            
            // Auto-refresh button positioned above status label
            autoRefreshButton.centerXAnchor.constraint(equalTo: view.centerXAnchor),
            autoRefreshButton.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor, constant: -80),
            autoRefreshButton.widthAnchor.constraint(equalToConstant: 200),
            autoRefreshButton.heightAnchor.constraint(equalToConstant: 44)
        ])
        
        // Don't auto-create MREC - wait for user to load
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        // Update status based on current ad state
        if mrecAd != nil && !isLoading {
            updateStatusUI(state: .ready)
        } else if isLoading {
            updateStatusUI(state: .loading)
        } else {
            updateStatusUI(state: .noAd)
        }
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        resetAdState()
    }
    
    deinit {
        // Ensure cleanup even if viewWillDisappear wasn't called
        resetAdState()
    }
    
    @objc private func loadMRECAd() {
        if isLoading {
            showAlert(title: "Info", message: "MREC is already loading.")
            return
        }
        
        resetAdState()
        createAndAddMRECToView()
        
        guard let mrecAd = mrecAd else { return }
        
        isLoading = true
        updateStatusUI(state: .loading)
        mrecAd.load()
    }
    
    private func createAndAddMRECToView() {
        guard mrecAd == nil else { return }
        
        var adUnitId = self.adUnitId
        if !settings.mrecAdUnitId.isEmpty {
            adUnitId = settings.mrecAdUnitId
        }
        mrecAd = CloudXCore.shared.createMREC(adUnitId: adUnitId)
        mrecAd?.delegate = self
        mrecAd?.revenueDelegate = self

        guard let mrecAd = mrecAd else {
            showAlert(title: "Error", message: "Failed to create MREC.")
            return
        }
        
        // Add MREC to view hierarchy immediately
        mrecAd.translatesAutoresizingMaskIntoConstraints = false
        view.addSubview(mrecAd)
        
        NSLayoutConstraint.activate([
            mrecAd.centerXAnchor.constraint(equalTo: view.centerXAnchor),
            mrecAd.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 210),
            mrecAd.widthAnchor.constraint(equalToConstant: 300),
            mrecAd.heightAnchor.constraint(equalToConstant: 250)
        ])
    }
    
    @objc private func loadMRECAdDeprecated() {
        if isLoading {
            showAlert(title: "Info", message: "MREC is already loading.")
            return
        }
        
        resetAdState()
        createAndAddMRECToViewDeprecated()
        
        guard let mrecAd = mrecAd else { return }
        
        isLoading = true
        updateStatusUI(state: .loading)
        mrecAd.load()
    }
    
    private func createAndAddMRECToViewDeprecated() {
        var adUnitId = self.adUnitId
        if !settings.mrecAdUnitId.isEmpty {
            adUnitId = settings.mrecAdUnitId
        }
        mrecAd = CloudXCore.shared.createMREC(adUnitId: adUnitId, viewController: self)
        mrecAd?.delegate = self
        mrecAd?.revenueDelegate = self
        
        guard let mrecAd = mrecAd else {
            showAlert(title: "Error", message: "Failed to create MREC (deprecated API).")
            return
        }
        
        mrecAd.translatesAutoresizingMaskIntoConstraints = false
        view.addSubview(mrecAd)
        
        NSLayoutConstraint.activate([
            mrecAd.centerXAnchor.constraint(equalTo: view.centerXAnchor),
            mrecAd.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 210),
            mrecAd.widthAnchor.constraint(equalToConstant: 300),
            mrecAd.heightAnchor.constraint(equalToConstant: 250)
        ])
    }
    
    // showMRECAd method removed - MREC is auto-added to view on push
    
    private func resetAdState() {
        if let mrecAd = mrecAd {
            // CRITICAL: Properly destroy the MREC to stop auto-refresh timers and background processing
            mrecAd.destroy()
            mrecAd.removeFromSuperview()
            self.mrecAd = nil
        }
        isLoading = false
    }
    
    @objc private func toggleAutoRefresh() {
        guard let mrecAd = mrecAd else {
            return
        }
        
        autoRefreshEnabled = !autoRefreshEnabled
        
        if autoRefreshEnabled {
            mrecAd.startAutoRefresh()
            autoRefreshButton.setTitle("Stop Auto-Refresh", for: .normal)
            autoRefreshButton.backgroundColor = .systemRed
        } else {
            mrecAd.stopAutoRefresh()
            autoRefreshButton.setTitle("Start Auto-Refresh", for: .normal)
            autoRefreshButton.backgroundColor = .systemGreen
        }
    }
    
    private var adUnitId: String {
        return CLXDemoConfigManager.sharedManager.currentConfig.mrecAdUnitId
    }
    
    private func loadMREC() {

        if isLoading || mrecAd != nil {
            return
        }

        isLoading = true
        updateStatusUI(state: .loading)

        let adUnitId = self.adUnitId
        mrecAd = CloudXCore.shared.createMREC(adUnitId: adUnitId)
        mrecAd?.delegate = self
        
        if let mrecAd = mrecAd {
            mrecAd.load()
        } else {
            isLoading = false
            updateStatusUI(state: .noAd)
            showAlert(title: "Error", message: "Failed to create MREC.")
        }
    }
    
    // MARK: - CLXBannerDelegate
    
    func didLoad(_ ad: CLXAd) {
        DemoAppLogger.sharedInstance.logAdEvent("✅ MREC didLoadAd", ad: ad)
        isLoading = false
        updateStatusUI(state: .ready)
        
        // Don't auto-show - user must press Show MREC button
    }
    
    func didFailToLoadAd(_ adUnitId: String, error: CLXError) {
        // No ad object exists on failure, so use logMessage instead of logAdEvent
        DemoAppLogger.sharedInstance.logMessage("❌ MREC failed to load for ad unit '\(adUnitId)' - Error: \(error.localizedDescription)")
        isLoading = false
        updateStatusUI(state: .noAd)
        
        DispatchQueue.main.async { [weak self] in
            let errorMessage = error.localizedDescription
            self?.showAlert(title: "MREC Error", message: errorMessage)
        }
    }
    
    func didClick(_ ad: CLXAd) {
        DemoAppLogger.sharedInstance.logAdEvent("👆 MREC didClickAd", ad: ad)
    }
    
    func didPayRevenue(for ad: CLXAd) {
        DemoAppLogger.sharedInstance.logAdEvent("💰 MREC didPayRevenue", ad: ad)
    }
    
    // Banner-specific delegate methods (MREC is a banner type)
    func didExpand(_ ad: CLXAd) {
        DemoAppLogger.sharedInstance.logAdEvent("🔍 MREC didExpandAd", ad: ad)
        
        DispatchQueue.main.async { [weak self] in
            self?.showAlert(title: "MREC Expanded!", 
                           message: "MREC ad expanded to full screen.")
        }
    }
    
    func didCollapse(_ ad: CLXAd) {
        DemoAppLogger.sharedInstance.logAdEvent("🔽 MREC didCollapseAd", ad: ad)
        
        DispatchQueue.main.async { [weak self] in
            self?.showAlert(title: "MREC Collapsed!", 
                           message: "MREC ad collapsed back to normal size.")
        }
    }
}